### ISSUE TITLE
/etc/cmdline.d/01-default.conf ignored

### ISSUE THREAD

**Issue by T35R6braPwgDJKq** [dracutdevs/dracut#1766]
**Describe the bug**
A generalization of #1754

**Distribution used**
Arch Linux

**Dracut version**
dracut 056

**Init system**
systemd

**To Reproduce**
```
sudo dracut --print-cmdline
 rd.luks.uuid=luks-SOMEUUID root=/dev/mapper/luks-SOMEOTHERUUID rootfstype=btrfs rootflags=rw,noatime,compress-force=zstd:3,space_cache=v2,subvolid=257,subvol=/@/0/snapshot

cat /proc/cmdline
BOOT_IMAGE=../vmlinuz-linux initrd=../initramfs-linux.img 

cat /etc/dracut.conf.d/dracut-defaults.conf
hostonly=yes
hostonly_cmdline=no
use_fstab=yes
compress=zstd
show_modules=yes
add_drivers+=' atkbd ' #lz4 lz4_compress
early_microcode=yes
CMDLINE=(
	rw
	rd.info
	rd.shell
	rd.luks.timeout=30
	rd.luks.crypttab=no
	rd.luks.uuid=SOMEUUID
	root=UUID=SOMEOTHERUUID
	rootflags=subvol=/@/0/snapshot
	zswap.enabled=1
	zswap.compressor=zstd
	zswap.zpool=z3fold
	rd.md=0
	rd.lvm=0
	rd.multipath=0
	rd.fcoe=0
	rd.neednet=1
	ip=dhcp
	resume=UUID=SOMEOTHERUUID
	resume_offset=533760
)
kernel_cmdline="${CMDLINE[*]}"
unset CMDLINE 
```

**Expected behavior**
After boot `/proc/cmdline` contains the cmdline defined in dracut, especially `resume` and `rd.luks.uuid`

**Additional context**
I would assume defining the `kernel_cmdline` in dracut relieves me from defining it again in my bootloader?!

**Comment by aafeijoo-suse** [comment]
I'd say you should set `hostonly_cmdline=yes`

**Comment by T35R6braPwgDJKq** [comment]
sounds good, doesnt work:

```
grep hostonly_cmdline /etc/dracut.conf.d/dracut-defaults.conf 
hostonly_cmdline=yes

dracut -f /boot/initramfs-linux.img

reboot

 cat /proc/cmdline 
BOOT_IMAGE=../vmlinuz-linux initrd=../initramfs-linux.img
```

**Comment by aafeijoo-suse** [comment]
Sorry, I misread your issue.
dracut custom cmdline options are not added to `/proc/cmdline`, but should be parsed by dracut at boot time.

**Comment by T35R6braPwgDJKq** [comment]
so it does not make sense to use things like `resume` or `resume_offset` in dracut? or does it have to be used twice? (dracut and bootloader?)
whats the purpose though? 

(btw. it does not seem to work as intended - e.g. `rd.luks` is ignored in dracut - see #1754 )

**Comment by dtardon** [comment]
> **Expected behavior** After boot `/proc/cmdline` contains the cmdline defined in dracut, especially `resume` and `rd.luks.uuid`

That's not how it works. I don't think it's even possible to change the command line of a running kernel. (`/proc/cmdline` is read-only for sure.) Instead, the functions used by dracut modules to get command line arguments (`getarg`, `getargs`, `getargbool`, `getargnum`) work on concatenation of the content of `/etc/cmdline.d/*.conf`, `/etc/cmdline` and `/proc/cmdline`.

> **Additional context** I would assume defining the `kernel_cmdline` in dracut relieves me from defining it again in my bootloader?!

It does, with one exception: these options wouldn't be visible for `ConditionKernelCommandLine=` directives in systemd units, because systemd (currently) only reads `/proc/cmdline`. Therefore, I suggest your problem actually lies somewhere else. E.g., if `resume` doesn't seem to be respected, is the `resume` module included in your initrd?

**Comment by T35R6braPwgDJKq** [comment]
Thanks for the support.
On another system with an UEFI unified image it worked that way (i.e. `/proc/cmdline` being the same as in dracut.conf) - so I assumed it should work with non-UEFI as well. A misunderstanding.
Your comment makes me feel even if `/proc/cmdline` differs from dracut.conf it should be respected in the running system (?)  - how would I check that?

how do I check if `resume` is respected and part of the initrd? I mean the kernel cmdline put out by `dracut -f ..` contains `resume` as specified. A naive `lsinitrd /boot/initramfs-linux.img| grep resume` does not yield anything though

**Comment by dtardon** [comment]
> Thanks for the support. On another system with an UEFI unified image it worked that way (i.e. `/proc/cmdline` being the same as in dracut.conf) - so I assumed it should work with non-UEFI as well. A misunderstanding.

Yeah, UEFI is very much different. In that case, dracut doesn't create normal initrd, but an UEFI executable (and collects all additional command line options to a separate file). IIUC, that executable is then executed by the boot loader, so these command line options can be passed to the kernel (while initrd is loaded by an already running kernel).

> Your comment makes me feel even if `/proc/cmdline` differs from dracut.conf it should be respected in the running system (?) - how would I check that?

I'm not sure I understand this. Could you rephrase it?

> 
> how do I check if `resume` is respected and part of the initrd? 

You're mixing two different things here: there is a kernel command line option named `resume` and a dracut module named `resume`, which reacts on the option. Resuming can only work if the `resume` module is included in the initrd *and* the `resume` command line option is used.

> I mean the kernel cmdline put out by `dracut -f ..` contains `resume` as specified. A naive `lsinitrd /boot/initramfs-linux.img| grep resume` does not yield anything though

There's your problem. `lsinitrd` lists the included modules; if `resume` isn't there, resuming cannot work, no matter if `resume` option is used. You may be hitting https://github.com/dracutdevs/dracut/issues/924. As a workaround, you can enforce inclusion of the module in config. file using `add_dracutmodules+=" resume "`.

> 
> > It does, with one exception: these options wouldn't be visible for ConditionKernelCommandLine= directives in systemd units,
> 
> Well, dracut fails to ask for a password if `rd.luks` is only specified in dracut.conf. It only asks for the password if it is specified in the bootloader. (thats the point of #1754 )

It's (a variation of) the exception I mentioned =) Unlocking of LUKS devices is handled via `cryptsetup-crypttab@.service` units generated from `/etc/crypttab` by `systemd-cryptsetup-generator`. But the generator only generates the units if it sees `luks` or `rd.luks` on the command line (and since systemd only reads `/proc/cmdline`... You see where I'm heading.) I'll comment on #1754 too.

Conclusion: the suspected generic problem described here doesn't exist, therefore the issue should be closed.

**Comment by T35R6braPwgDJKq** [comment]
I added the `resume` module now and it is part of the initrd (wouldnt it make sense to add that automatically, if the `kernel_cmdline` already contains a `resume=`?). But i'm still not able to hibernate:
```
 /usr/lib/systemd/systemd-sleep hibernate
Failed to find location to hibernate to: Function not implemented
```
What point is there though to define `rd.luks` anyway in dracut if it does not work except for a unified kernel image?

**Comment by dtardon** [comment]
> I added the `resume` module now and it is part of the initrd (wouldnt it make sense to add that automatically, if the `kernel_cmdline` already contains a `resume=`?).

Maybe. 

> But i'm still not able to hibernate:
> 
> ```
>  /usr/lib/systemd/systemd-sleep hibernate
> Failed to find location to hibernate to: Function not implemented
> ```

Try to run with `SYSTEMD_LOG_LEVEL=debug` exported. It should give more info in the log.
 
> What point is there though to define `rd.luks` anyway in dracut if it does not work except for a unified kernel image?

Presumably it also works if systemd is not used as init... dracut had added support for /etc/cmdline before systemd even existed. It's not dracut's fault that systemd doesn't want to read that too.

Anyway, I've been thinking about pure dracut workaround, but I don't think there really is one. The only thing I come up with that might possibly work is to have a service that would write the extended command line options to a file, bind mount it over `/proc/cmdline`, and call `systemctl daemon-reload` (so generators are re-run). It would also unmount it on stop again (to not clutter `/proc/cmdline` of the system with initrd options). But it is racy unless this service runs before every other service. (It's also ugly, but that's a different matter.)
